dashboard/src/components/TestDetails/TestDetails.tsx

```diff
@@ -22,22 +28,42 @@ import IssueSection from '../Issue/IssueSection';
 const emptyValue = '-';
 const valueOrEmpty = (value: string | undefined): string => value || emptyValue;
 
-const TestDetailsSection = ({
-  test,
-  context,
-}: {
-  test: TTestDetails;
-  context: string;
-}): JSX.Element => {
+const TestDetailsSection = ({ test }: { test: TTestDetails }): JSX.Element => {
   const intl = useIntl();
+  const historyState = useRouterState({ select: s => s.location.state });
+  const searchParams = useSearch({ from: '/test/$testId' });
   const hardware: string =
     test.environment_compatible?.join(' | ') ??
     intl.formatMessage({ id: 'global.unknown' });
 
-  const buildDetailsLink =
-    `${window.location.origin}` +
-    `/${context}/${test.git_commit_hash}/build/${test.build_id}` +
-    `${window.location.search}`;
+  const buildDetailsLink = useMemo(() => {
+    let linkTo = '';
+    let linkParams = {};
+    if (historyState.from === RedirectFrom.Hardware && historyState.id) {
+      linkTo = '/hardware/$hardwareId/build/$buildId';
+      linkParams = { hardwareId: historyState.id, buildId: test.build_id };
+    } else if (historyState.from === RedirectFrom.Tree && historyState.id) {
+      linkTo = '/tree/$treeId/build/$buildId';
+      linkParams = { treeId: historyState.id, buildId: test.build_id };
+    } else {
+      linkTo = '/build/$buildId';
+      linkParams = { buildId: test.build_id };
+    }
+
+    return (
+      <Link
+        to={linkTo}
+        params={linkParams}
+        search={searchParams}
+        className="flex flex-row items-center gap-1"
+        target="_blank"
+        rel="noreferrer"
+      >
+        {truncateBigText(test.build_id)}
+        <FiLink className="text-blue" />
+      </Link>
+    );
+  }, [historyState, test.build_id, searchParams]);
```

Comment:
Why you don't use `LinkWithIcon` component and pass `linkText=truncateBigText(test.build_id)` and `icon=<FiLink className="text-blue" />`?

Reply:
As stated by Wilson in a previous comment, it would be better to use tanstack router to get the search params. LinkWithIcon does not accept the object returned by useSearch as a prop (since LinkWithIcon is basically an 'a' HTML element) so the solution was to use tanstack router `Link` and modify the LinkWithIcon component to render this Link in case it's passed to it instead of 'a' (or 'div')

Reply:
I was originally using LinkWithIcon but that would mean I'd have to build the link using things like `location.search` for example